extensions/DynamicPageList/DplRevisionQuerySegmentBuilder.php

```diff
@@ -0,0 +1,186 @@
+		} else {
+			$userId = $this->databaseConnection->addQuotes( $userArray['user_id'] );
+			$sqlQuerySegment = "AND change_rev.rev_user = $userId";
+		}
+
+		$sqlQuerySegment .= " AND change_rev.rev_page = page_id";
+
+		return " $sqlQuerySegment";
+	}
+
+	/**
+	 * Construct an SQL query segment that will match only revisions NOT modified by the given user
+	 * @param array $userArray An array with either an user_text entry or an user_id entry
+	 * @return string the SQL query segment
+	 */
+	public function buildNotModifiedByQuerySegment( array $userArray ): string {
+		if ( empty( $userArray ) ) {
+			return '';
+		}
+
+		$revisionTable = $this->databaseConnection->tableName( 'revision' );
+
+		if ( isset( $userArray['user_text'] ) ) {
+			$userText = $this->databaseConnection->addQuotes( $userArray['user_text'] );
+			$whereStatement = "$revisionTable.rev_user_text = $userText";
+		} else {
+			$userId = $this->databaseConnection->addQuotes( $userArray['user_id'] );
+			$whereStatement = "$revisionTable.rev_user = $userId";
+		}
+
+		$sqlQuerySegment = <<<SQL
+ AND NOT EXISTS (SELECT 1 FROM $revisionTable WHERE $revisionTable.rev_page=page_id AND $whereStatement LIMIT 1)
+SQL;
+
+		return $sqlQuerySegment;
+	}
+
+	/**
+	 * Construct an SQL query segment that will match only revisions last modified by the given user
+	 * @param array $userArray An array with either an user_text entry or an user_id entry
+	 * @return string the SQL query segment
+	 */
+	public function buildLastModifiedByQuerySegment( array $userArray ): string {
+		if ( empty( $userArray ) ) {
+			return '';
+		}
+
+		$revisionTable = $this->databaseConnection->tableName( 'revision' );
+
+		if ( isset( $userArray['user_text'] ) ) {
+			$userNameOrId = $this->databaseConnection->addQuotes( $userArray['user_text'] );
+			$userField = 'rev_user_text';
+		} else {
+			$userNameOrId = $this->databaseConnection->addQuotes( $userArray['user_id'] );
+			$userField = 'rev_user';
+		}
+
+		$sqlQuerySegment = <<<SQL
+ AND (SELECT $userField FROM $revisionTable WHERE $revisionTable.rev_page=page_id ORDER BY $revisionTable.rev_timestamp DESC LIMIT 1) = $userNameOrId
+SQL;
+
+		return $sqlQuerySegment;
+	}
+
+	/**
+	 * Construct an SQL query segment that will match only revisions NOT last modified by the given user
+	 * @param array $userArray An array with either an user_text entry or an user_id entry
+	 * @return string the SQL query segment
+	 */
+	public function buildNotLastModifiedByQuerySegment( array $userArray ): string {
+		if ( empty( $userArray ) ) {
+			return '';
+		}
+
+		$revisionTable = $this->databaseConnection->tableName( 'revision' );
+
+		if ( isset( $userArray['user_text'] ) ) {
+			$userNameOrId = $this->databaseConnection->addQuotes( $userArray['user_text'] );
+			$userField = 'rev_user_text';
+		} else {
+			$userNameOrId = $this->databaseConnection->addQuotes( $userArray['user_id'] );
+			$userField = 'rev_user';
+		}
+
+		$sqlQuerySegment = <<<SQL
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Immediately return this expression instead of assigning it to the temporary variable "$sqlQuerySegment". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1488)

[project-id]: wikia:app